### PR TITLE
Decouple HeuristicConfigPolicy and ScoreDirectorFactory plus related tests cleanup

### PR DESCRIPTION
**Pull request by yurloc**
Benefits:
- ScoreDirectorFactory cannot leak from the policy object.
- Callers looking for the InitializingScoreTrend do not have to traverse through the SDF.
- Easier to create a HeuristicConfigPolicy instance because SDF is no longer needed (tests do not have to create an SDF mock).

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] tests</b>
  
- for a <b>full downstream build</b> 
  - for <b>jenkins</b> job: 
    please add comment: <b>Jenkins run fdb</b>
  - for <b>github actions</b> job: 
    add the label `run_fdb`

- for a <b>compile downstream build</b>  
  please add comment: <b>Jenkins run cdb</b>

- for a <b>full production downstream build</b>  
  please add comment: <b>Jenkins execute product fdb</b>

- for an <b>upstream build</b>  
  please add comment: <b>Jenkins run upstream</b>

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] quarkus-branch</b>

- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] native</b>

- for <b>mandrel checks</b>  
  Run native checks against Mandrel image
  Please add comment: <b>Jenkins run mandrel</b>

- for a <b>specific mandrel check</b>  
  Run native checks against Mandrel image  
  Please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] mandrel</b>
</details>

### CI Status

 You can check OptaPlanner repositories CI status from [Chain Status webpage](https://kiegroup.github.io/optaplanner/).

### CODE REVIEW

File: core/optaplanner-core-impl/src/test/java/org/optaplanner/core/impl/constructionheuristic/placer/entity/PlacementAssertions.java
Comment (triceo) [inline core/optaplanner-core-impl/src/test/java/org/optaplanner/core/impl/constructionheuristic/placer/entity/PlacementAssertions.java:12]: I'm not too keen on deleting these abstract parent classes.

I know that they've been mostly empty, but sharing code is not their only value - how about code structure? Until this PR, we know what the selector factory tests were, what the score definition tests were, etc. And now, they all blend into the crowd.

I am not going to block the PR for this, as I really can't tell if this is important or not, but I do wonder if we're losing something here.

**Review by triceo — APPROVED** [review]
LGTM, but see the comment inside.

File: core/optaplanner-core-impl/src/test/java/org/optaplanner/core/impl/constructionheuristic/placer/entity/PlacementAssertions.java
Comment (yurloc) [reply core/optaplanner-core-impl/src/test/java/org/optaplanner/core/impl/constructionheuristic/placer/entity/PlacementAssertions.java:12]: You have a valid point and after thinking about it carefully and doing some investigation, I disagree.

> Until this PR, we know what the selector factory tests were, what the score definition tests were, etc.

We still know it even after this PR.
- The score definition tests are all in the `org.optaplanner.core.impl.score.buildin` package.
- The entity/value placer tests are _now_ all in the `org.optaplanner.core.impl.constructionheuristic.placer.entity` package.
- The selector factory tests *not in the same pacakge* but they are all named `*SelectorFactoryTest`.

Good naming practices and following the package structure laid out by the `src/main/java` code are the right tools to organize our tests in my opinion.

The removed abstract classes did not help with the code organization because this practice was not applied consistently:
- The `PooledEntityPlacerFactoryTest` did not extend the `AbstractEntityPlacerTest`. (I found it while deduplicating the `buildHeuristicConfigPolicy()` method.)
- The `HardSoftScoreDefinitionTest` was the only child of the `AbstractScoreDefinitionTest`.
- There are selector factory tests that did not extend the `AbstractSelectorFactoryTest`.

So, in my opinion, using abstract classes to tag similar tests is not useful and we don't lose anything by removing them.

I'll leave some time for @ge0ffrey to share his opinion and I'll merge this after that.

Comment (ge0ffrey) [reply core/optaplanner-core-impl/src/test/java/org/optaplanner/core/impl/constructionheuristic/placer/entity/PlacementAssertions.java:12]: No strong opinion from my side.

If it empty now (its builder methods are better at another place?), I guess its natural to delete it?

**Pull request merged**
